vuln-reach/src/test_util.rs

```diff
@@ -0,0 +1,60 @@
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use crate::javascript::module::TarballModuleResolver;
+use crate::javascript::package::Package;
+
+/// Retrieve the tarball bytes for a given package and version. Will use the
+/// on-disk cache or download the package from npm.
+pub async fn get_tarball_bytes(package_name: &str, version: &str) -> Vec<u8> {
+    let package_path = tarball_path(package_name, version);
+
+    if package_path.exists() {
+        fs::read(package_path).expect("Could not read tarball contents")
+    } else {
+        download_tarball(package_name, version, &package_path).await
+    }
+}
+
+pub async fn npm_package(name: &str, version: &str) -> Package<TarballModuleResolver> {
+    Package::from_tarball_bytes(get_tarball_bytes(name, version).await).unwrap()
+}
+
+// Return the path to the tarball cache. Create the directory if it doesn't
+// exist.
+fn tarball_cache_path() -> PathBuf {
+    let tarball_cache_path = PathBuf::from(env!("OUT_DIR")).join("tarball_cache");
+    if !tarball_cache_path.exists() {
+        fs::create_dir_all(&tarball_cache_path).expect("Could not create tarball cache directory");
+    }
+
+    tarball_cache_path
+}
+
+fn tarball_path(package_name: &str, version: &str) -> PathBuf {
+    tarball_cache_path().join(format!("{package_name}-{version}.tgz"))
+}
+
+async fn download_tarball(full_package_name: &str, version: &str, package_path: &Path) -> Vec<u8> {
+    // Turn "@foo/bar" into ("@foo/bar", "bar") and "foo" into ("foo", "foo").
+    let (namespace, package_name) = full_package_name
+        .split_once('/')
+        .map(|(_, b)| (full_package_name, b))
+        .unwrap_or_else(|| (full_package_name, full_package_name));
+
+    let package_uri =
+        format!("https://registry.npmjs.org/{namespace}/-/{package_name}-{version}.tgz");
```

Comment:
The stability of this URI format is a bold assumption. I think it should be fine, though: we control all of its inputs at test time and so we don't need to find a solution which is completely generic.

Reply:
URI stability is pretty much irrelevant for tests. This would be trivial to fix once changed and I don't expect this to be changing much (if ever).